### PR TITLE
No Bug: Fixing DAU ping race condition on app activation.

### PR DESCRIPTION
**Pull request by jhreis**
I need to build out some better unit tests for this, however some larger architecture stuff will probably need to be adjusted a bit.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.

## Reviewer Checklist:

- [x] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [x] Issues are assigned to at least one epic.
- [x] Issues include necessary QA labels:
  - [x] `QA/(Yes|No)`
  - [x] `release-notes/(include|exclude)`
  - [x] `bug` / `enhancement`
- [x] Necessary security reviews have taken place.
- [x] Adequate test coverage exists to prevent regressions.
- [x] Adaquate test plan exists for QA to validate (if applicable)

### CODE REVIEW

**Review by iccub — APPROVED** [review]

File: BraveShared/Analytics/DAU.swift [hunk 1]
```diff
@@ -38,6 +40,11 @@ public class DAU {
             return false
         }
         
+        if processingPing {
```
Comment (danishjafri88) [inline BraveShared/Analytics/DAU.swift]: This is redundant once you make the DAU class object global. The previous check also do the same thing. Basically you cant have a sync check for a async updation. Also the check below for launchTimer != nil will make sure no new timers are created.

Comment (danishjafri88) [inline BraveShared/Analytics/DAU.swift]: This is redundant once you make the DAU class object global. The check below for launchTimer != nil will make sure no new timers are created. Instead put this check at `guard let paramsAndPrefs = paramsAndPrefsSetup() else ` and set `processingPing = true` after the guard

File: BraveShared/Analytics/DAU.swift [hunk 2]
```diff
@@ -56,6 +63,7 @@ public class DAU {
     }
     
     @objc public func sendPingToServerInternal() {
+        processingPing = true
         guard let paramsAndPrefs = paramsAndPrefsSetup() else {
```
Comment (danishjafri88) [inline BraveShared/Analytics/DAU.swift]: Check for `processingPing ` also and set `processingPing` as true after the guard.

File: BraveShared/Analytics/DAU.swift [hunk 3]
```diff
@@ -86,6 +94,7 @@ public class DAU {
             Preferences.DAU.lastLaunchInfo.value = paramsAndPrefs.lastLaunchInfoPreference
             
             Preferences.DAU.lastPingFirstMonday.value = paramsAndPrefs.lastPingFirstMondayPreference
+            self.processingPing = false
```
Comment (danishjafri88) [inline BraveShared/Analytics/DAU.swift]: You need to set self.processingPing = false in above `if let e = error {` also to complete the check.

**Review by danishjafri88 — CHANGES_REQUESTED** [review]

**Comment by danishjafri88** [comment]
The change is required, let take a case:
Time fires 1st time: Calls the `sendPingToServerInternal `  and `processingPing ` is set as true.
before the api call is done (Should not if the time is more than api timeout but for architecture sake). The timer fires again and there is no check to make sure the api call is not done.

**Comment by jhreis** [comment]
@danishjafri88 yeah, good point, will adjust.

File: BraveShared/Analytics/DAU.swift (hunk 2 above)
Comment (jhreis) [reply BraveShared/Analytics/DAU.swift]: Added the former, but why the latter?

File: BraveShared/Analytics/DAU.swift (hunk 1 above)
Comment (jhreis) [reply BraveShared/Analytics/DAU.swift]: Did move, and `dau.sendPingToServer()` is still called manually on app open.

File: BraveShared/Analytics/DAU.swift (hunk 3 above)
Comment (jhreis) [reply BraveShared/Analytics/DAU.swift]: yikes, thx

File: BraveShared/Analytics/DAU.swift (hunk 2 above)
Comment (danishjafri88) [reply BraveShared/Analytics/DAU.swift]: It will always be processingPing ie true if paramsAndPrefsSetup failed for today but not after midnight

**Review by danishjafri88 — APPROVED** [review]

**Pull request merged**